### DISCUSSION TITLE
asyncio.exceptions.CancelledError not being handled properly inside Docker image

### DISCUSSION THREAD

**Discussion by jzr-supove** [Kludex/uvicorn#2017] [Potential Issue] [answered]
`asyncio.exceptions.CancelledError` not being handled properly in uvicorn app running inside Docker image:

```
INFO:     Started server process [1]      
INFO:     Waiting for application startup.
INFO:     Application startup complete.                                
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
^CINFO:     Shutting down
INFO:     Finished server process [1]
ERROR:    Traceback (most recent call last):                           
  File "***/starlette/routing.py", line 686, in lifespan 
  File "***/uvicorn/lifespan/on.py", line 139, in receive
  File "***/asyncio/queues.py", line 159, in get
asyncio.exceptions.CancelledError
```

I'm trying to kill the running process if it doesn't pass the validation at the startup event (FastAPI). This is the function that is being executed when there is some Exception occurs during the startup event:

```python3
def exit_program():
    parent_pid = os.getpid()
    parent = psutil.Process(parent_pid)
    for child in parent.children(recursive=True):
        child.kill()
    parent.kill()
```

Here I'm using `psutil` package for killing the process and its children. Because I couldn't find any straightforward approach (api) to shutdown the uvicorn server programmatically and exit the script (If there is one, please guide)
**I think there should be some convenient api for this, like `uvicorn.shutdown(app)`, or even better uvicorn should properly handle the exit() function and shutdown itself gracefully (or forcefully, doesn't matter)**

On local, its working without any errors, but when I try building a docker image and run it, program just hangs, I need to send SIGINT signal by pressing Ctrl+C, then the `asyncio.exceptions.CancelledError` pops up, along with the logs of my exit function.

Just in case, here is full demo:
```python3
import os
import psutil
import uvicorn
from fastapi import FastAPI
from apscheduler.schedulers.asyncio import AsyncIOScheduler


def exit_program():
    parent_pid = os.getpid()
    parent = psutil.Process(parent_pid)
    for child in parent.children(recursive=True):
        child.kill()
    parent.kill()


scheduler = AsyncIOScheduler()

def validate(start=False):
    try:
        1/ 0  # Some validation process
    except Exception:
        print("Error")
        exit_program()

def schedule_background_task():
    scheduler.add_job(validate, 'interval', minutes=10)
    scheduler.start()

app = FastAPI(title='a', docs_url=None, redoc_url=None)

@app.get("/")
async def index():
    return True

@app.on_event("startup")
async def startup_event():
    validate(start=True)
    schedule_background_task()

if __name__=='__main__':
    uvicorn.run(app, host="0.0.0.0", port=8000)
```

**Comment by jzr-supove** [comment]
I don't understand why you are just taking and converting the issue to discussion, and closing it without saying anything...
**Issue is still not resolved**

**Reply by Kludex** [reply]
The starting point should be a discussion, as mention in the issue template:

<img width="1033" alt="image" src="https://github.com/encode/uvicorn/assets/7353520/59c8fb5e-a189-49f9-8c58-bb7729df1ec0">

**Reply by jzr-supove** [reply]
Oh I see

**Comment by jzr-supove** [comment, marked as answer]
After hours of debugging and trying different things, I finally managed to solve the issue.

I changed `CMD ["command"]` syntanx to `CMD "command"` (without brackets) inside `Dockerfile`, and it started working correctly.
I don't know if it is a right approach, but I'm happy that its working. Seems like it was a Docker issue, not an issue related with Python or Uvicorn. Sorry for bothering

Anyways, it still would be great if you actually implement some convenient method for shutting down the server programmatically, like `uvicorn.shutdown(app)` or something like that. Thanks <3

**Reply by Kludex** [reply]
Same reply as https://github.com/encode/uvicorn/issues/1215#issuecomment-939372515 then.
